src/main/scala/com.snowplowanalytics.sauna/utils/Command.scala

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (c) 2017 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.sauna
+package utils
+
+// java
+import java.time.LocalDateTime
+import java.time.temporal.ChronoUnit
+
+// scalaz
+import scalaz.{Failure, Success}
+
+// play
+import play.api.libs.functional.syntax._
+import play.api.libs.json._
+
+// jsonschema
+import com.fasterxml.jackson.databind.JsonNode
+import com.github.fge.jsonschema.main.JsonSchemaFactory
+
+// iglu
+import com.snowplowanalytics.iglu.client.repositories.{HttpRepositoryRef, RepositoryRefConfig}
+import com.snowplowanalytics.iglu.client.{SchemaKey, _}
+
+object Command {
+  /**
+   * Self-describing data.
+   *
+   * @param schema The schema the data should be validated against.
+   * @param data   The data itself.
+   */
+  case class SelfDescribing(
+    schema: SchemaKey,
+    data: JsValue
+  )
+
+  /**
+   * A Sauna command.
+   *
+   * @param envelope Identifies the command and provides instructions on how to execute it.
+   * @param command  Specifies the exact action to execute.
+   */
+  case class SaunaCommand(
+    envelope: SelfDescribing,
+    command: SelfDescribing)
+
+  implicit val schemaKeyReads: Reads[SchemaKey] = Reads {
+    case JsString(s) => SchemaKey.parse(s) match {
+      case Success(key) => JsSuccess(key)
+      case Failure(error) => JsError(error.toString)
+    }
+    case _ => JsError("Non-string SchemaKey")
+  }
+
+  implicit val selfDescribingDataReads: Reads[SelfDescribing] = (
+    (JsPath \ "schema").read[SchemaKey] and
+      (JsPath \ "data").read[JsValue]
+    ) (SelfDescribing.apply _)
+
+  implicit val saunaCommandReads: Reads[SaunaCommand] = (
+    (JsPath \ "envelope").read[SelfDescribing] and
+      (JsPath \ "command").read[SelfDescribing]
+    ) (SaunaCommand.apply _)
+
+  sealed trait Semantics
+  case object AT_LEAST_ONCE extends Semantics
+
+  implicit val semanticsReads: Reads[Semantics] = Reads {
+    case JsString(s) => s match {
+      case "AT_LEAST_ONCE" => JsSuccess(AT_LEAST_ONCE)
+      case _ => JsError(s"Invalid Semantics: $s")
+    }
+    case _ => JsError("Non-string Semantics")
+  }
+
+  /**
+   * TODO: define what this is
```

Comment:
@alexanderdean: can you extend the [general command spec](https://github.com/snowplow/sauna/wiki/Commands-for-analysts) with information about the currently unused metadata? (`semantics`, `tags` etc.) I'll incorporate the info into these docs when that's done.